src/lib.rs

```diff
@@ -218,44 +259,120 @@ impl<T> Sender<T> {
     /// the error involves running any drop implementation on the message type, which might or
     /// might not be lock-free.
     pub fn send(self, message: T) -> Result<(), SendError<T>> {
-        // SAFETY: The channel exists on the heap for the entire duration of this method.
-        let channel: &mut Channel<T> = unsafe { &mut *self.channel_ptr };
+        let channel_ptr = self.channel_ptr;
 
         // Don't run our Drop implementation if send was called, any cleanup now happens here
         mem::forget(self);
 
+        // SAFETY: The channel exists on the heap for the entire duration of this method and we
+        // only ever acquire shared references to it. Note that if the receiver disconnects it
+        // does not free the channel.
+        let channel = unsafe { channel_ptr.as_ref() };
+
         // Write the message into the channel on the heap.
-        channel.write_message(message);
+        // SAFETY: The receiver only ever accesses this memory location if we are in the MESSAGE
+        // state, and since we're responsible for setting that state, we can guarantee that we have
+        // exclusive access to this memory location to perform this write.
+        unsafe {
+            channel.write_message(message);
+        }
```

Comment:
If you put the semicolon after the closing bracket, it becomes a single line of code. This is a neat trick that both of us have already used a lot.